src/wasm/wasm-type.cpp

```diff
@@ -2301,9 +2312,122 @@ template<> class hash<wasm::ShallowHeapType> {
 namespace wasm {
 namespace {
 
-// Uses Hopcroft's DFA minimization algorithm to construct a minimal type
-// definition graph from an input graph. See
-// https://en.wikipedia.org/wiki/DFA_minimization#Hopcroft's_algorithm.
+// The Refined Partitions data structure used in Valmari-Lehtinen DFA
+// minimization.
+struct Partitions {
+  // The number of sets.
+  size_t sets = 0;
+
+  // The partitioned elements. Elements in the same set are next to each other.
+  // Within each set, "marked" elements come first followed by "unmarked"
+  // elements.
+  std::vector<size_t> elements;
+
+  // Maps elements to their indices in `elements`.
+  std::vector<size_t> elementIndices;
+
+  // Maps elements to their sets, identified by an index.
+  std::vector<size_t> setIndices;
+
+  // Maps sets to the indices of their first elements in `elements`.
+  std::vector<size_t> beginnings;
+
+  // Maps sets to (one past) the indices of their ends in `elements`.
+  std::vector<size_t> endings;
+
+  // Maps sets to the indices of their first unmarked elements in `elements`.
+  std::vector<size_t> pivots;
```

Comment:
Is not better group `elementIndices` and `setIndices` into AoS? like `vector<{ elementIndex, setIndicex }>`. And same for `beginnings`, `endings` and `pivots` -> `vector<{ beginning, ending, pivot }>`.

Reply:
Possibly; I haven't investigated that yet but I was thinking about it. The reduced memory allocations shouldn't matter because all the allocations happen just once during initialization, but I would be interested to see whether cache locality would be better if we used arrays of structs for more parts. For this PR, I'm happy just sticking to what the paper describes, though.